### PR TITLE
`Chore`: Re-update to serialization based navigation

### PR DESCRIPTION
**Pull request by FelberMartin**
### Problem Description
Tim opened https://github.com/ls1intum/artemis-android/pull/49 that introduced type safe navigation. After prematurely merging the PR, I found that the changes cased a crash. See https://github.com/ls1intum/artemis-android/pull/49#issuecomment-2433172761.

This closes issue #104.

### Changes
 
Readded the changes that Tim already did, and fix two crash causing bugs.

### Steps for testing
Navigate through the app and ensure that everything works as expected. Especially cover the case that the user is not logged in when opening the app.

Cover navigation to the following screens:
- Login
- Registration (eg TS2)
- Dashboard
- ExerciseView
  - Quiz
- LectureScreen
- CommunicationOverview
- Settings
  - PushNotificationsSettings

### CODE REVIEW

**Review by julian-wls — APPROVED** [review]
I tested it by navigating to every possible screen and didn't experience any issues. Everything worked as expected.
Code lgtm!

**Comment by FelberMartin** [comment]
Due to a bug in the RobolectricTestRunner I had to ignore the tests using a ModalBottomSheet. 

This closed dependency update can also be reopened with the ignored tests: https://github.com/ls1intum/artemis-android/pull/46

File: app/src/main/java/de/tum/informatics/www1/artemis/native_app/android/ui/MainActivity.kt [hunk 1]
```diff
@@ -347,7 +350,7 @@ class MainActivity : AppCompatActivity(),
                 quizParticipation(
                     onLeaveQuiz = {
                         val previousBackStackEntry = navController.previousBackStackEntry
-                        if (previousBackStackEntry?.destination?.route == ExerciseViewDestination.EXERCISE_VIEW_ROUTE) {
+                        if (previousBackStackEntry?.destination?.route == ExerciseViewUi::class.qualifiedName.orEmpty()) {
```
Comment (TimOrtel) [inline app/src/main/java/de/tum/informatics/www1/artemis/native_app/android/ui/MainActivity.kt:353]: Did you verify test this code?

Comment (FelberMartin) [reply app/src/main/java/de/tum/informatics/www1/artemis/native_app/android/ui/MainActivity.kt:353]: We have currently disabled Quiz participation, see https://github.com/ls1intum/artemis-android/issues/107. I will add this check as a todo there

**Pull request merged**